handler/oauth2/flow_authorize_jwt_token.go

```diff
@@ -0,0 +1,330 @@
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant \"%s\".", grantTypeJwtBearer))
+	}
+
+	assertion := request.GetRequestForm().Get("assertion")
+	if assertion == "" {
+		return errorsx.WithStack(fosite.ErrInvalidRequest.WithHintf("The assertion request parameter must be set when using grant_type of '%s'.", grantTypeJwtBearer))
+	}
+
+	token, err := jwt.ParseSigned(assertion)
+	if err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Unable to parse jwt token passed in \"assertion\" request parameter.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+
+	// check fo required claims in token, so we can later find public key based on them
+	if err := c.validateTokenPreRequisites(token); err != nil {
+		return err
+	}
+
+	key, err := c.findPublicKeyForToken(ctx, token)
+	if err != nil {
+		return err
+	}
+
+	if err := c.validateToken(ctx, token, key); err != nil {
+		return err
+	}
+
+	claims := jwt.Claims{}
+	if err := token.Claims(key, &claims); err != nil {
+		return errorsx.WithStack(fosite.ErrInvalidGrant.
+			WithHint("Unable to verify the integrity of the 'assertion' value.").
+			WithWrap(err).WithDebug(err.Error()),
+		)
+	}
+
+	if claims.ID != "" {
+		if err := c.AuthorizeJwtGrantStorage.MarkJWTUsedForTime(ctx, claims.ID, claims.Expiry.Time()); err != nil {
+			return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+		}
+	}
+
+	scopes, err := c.AuthorizeJwtGrantStorage.GetPublicKeyScopes(ctx, claims.Issuer, claims.Subject, key.KeyID)
+	if err != nil {
+		return errorsx.WithStack(fosite.ErrServerError.WithWrap(err).WithDebug(err.Error()))
+	}
+
+	for _, scope := range request.GetRequestedScopes() {
+		if !c.ScopeStrategy(scopes, scope) {
+			return errorsx.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope '%s'.", scope))
+		}
+	}
+
+	for _, scope := range request.GetRequestedScopes() {
+		request.GrantScope(scope)
+	}
+
+	for _, audience := range claims.Audience {
+		request.GrantAudience(audience)
+	}
+
+	session, err := c.getSessionFromRequest(request)
+	if err != nil {
+		return err
+	}
+	session.SetExpiresAt(fosite.AccessToken, time.Now().UTC().Add(c.HandleHelper.AccessTokenLifespan).Round(time.Second))
+	session.SetSubject(claims.Subject)
+
+	return nil
+}
+
+func (c *AuthorizeJwtGrantHandler) PopulateTokenEndpointResponse(ctx context.Context, request fosite.AccessRequester, response fosite.AccessResponder) error {
+	if !c.CanHandleTokenEndpointRequest(request) {
+		return errorsx.WithStack(fosite.ErrUnknownRequest)
+	}
+
+	// if client is authenticated, check grant types
+	if !c.CanSkipClientAuth(request) && !request.GetClient().GetGrantTypes().Has(grantTypeJwtBearer) {
+		return errorsx.WithStack(fosite.ErrUnauthorizedClient.WithHintf("The OAuth 2.0 Client is not allowed to use authorization grant \"%s\".", grantTypeJwtBearer))
+	}
```

Comment:
This is duplicated from above - maybe it can be extracted into a common method?

Reply:
yes, why not)